component/distributed_cache/distributed_cache.go

```diff
@@ -356,68 +361,265 @@ func (dc *DistributedCache) StreamDir(options internal.StreamDirOptions) ([]*int
+			log.Err("DistributedCache::CreateFile : File Creation failed with err : %s Dcache subcomponent, path : %s", err.Error(), options.Name)
 			return nil, err
 		}
 
-		_, err = dc.NextComponent().CreateFile(options)
+		handle, err = dc.NextComponent().CreateFile(options)
 		if err != nil {
+			log.Err("DistributedCache::CreateFile : File Creation failed with err : %s Azure subcomponent, path : %s", err.Error(), options.Name)
 			return nil, err
 		}
 		// todo : if one is success and other is failure, get to the previous state by removing the
 		// created entries for the files.
 	}
 
-	handle := handlemap.NewHandle(options.Name)
+	if handle == nil {
+		handle = handlemap.NewHandle(options.Name)
+	}
+
+	// Set the respective filesystems that this path can access
+	if isAzurePath {
+		handle.SetFsAzure()
+	} else if isDcachePath {
+		handle.SetFsDcache()
+	} else {
+		handle.SetFsDefault()
+	}
+
+	// Set Dcache file inside the handle
+	handle.IFObj = dcFile
 
-	handle.IFObj = file
+	// Only the files created with dcache needs flush.
+	handle.Flags.Set(handlemap.HandleFlagDcacheAllowWrites)
 
 	return handle, nil
 }
 func (dc *DistributedCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
 	if options.Flags&os.O_WRONLY != 0 || options.Flags&os.O_RDWR != 0 {
-		log.Info("DistributedCache::OpenFile: Writing to an existing File is not allowed, file : %s", options.Name)
+		log.Err("DistributedCache::OpenFile: Writing to an exisiting File is not allowed, file : %s", options.Name)
 		return nil, syscall.EACCES
 	}
 
-	return nil, syscall.ENOTSUP
+	var dcFile *fm.DcacheFile
+	var handle *handlemap.Handle
+	var err error
+
+	isAzurePath, isDcachePath, rawPath := getFS(options.Name)
+	if isAzurePath {
+		log.Debug("DistributedCache::OpenFile : Path is having Azure subcomponent, path : %s", options.Name)
+		options.Name = rawPath
+		handle, err = dc.NextComponent().OpenFile(options)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Azure subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+	} else if isDcachePath {
+		log.Debug("DistributedCache::OpenFile : Path is having Dcache subcomponent, path : %s", options.Name)
+		options.Name = rawPath
+		dcFile, err = fm.OpenDcacheFile(options.Name)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Dcache subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+	} else {
+		common.Assert(options.Name == rawPath)
+		dcFile, err = fm.OpenDcacheFile(rawPath)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Dcache subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+		handle, err = dc.NextComponent().OpenFile(options)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Azure subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
```

Comment:
opening file for read with no explicit namespace specified, must have the following semantics.
It should first check the file in dcache, if present, read from dcache, else check in azure if present, read from azure, else fail the open.
So we should do this here

open in dcache
if it succeeds, call SetFsDcache() and set handle.IFObj (to indicate this file must be read from dcache) and return.
if this fails with enoent then try to open in azure, else if it fails with any other error, fail the open
if azure open fails, fail the open call, else call SetFsAzure() to indicate this file must be read from azure.